### ISSUE TITLE
Dynamic label rendering for multiple cylinder shapes

### ISSUE THREAD

**Issue by JavierSanchez-Utges** [3dmol/3Dmol.js#803]
Good afternoon,

I am displaying multiple cylinder shapes that I want to annotate with labels on hover and loose the labels when hovering out. I have it working for one label, that is the label for one cylinder which arguments are hard coded in the hover callback function:

```
var ligandLabel = null; // Shared variable for the label
var interLabel = null; // Shared variable for the label
var protLabel = null; // Shared variable for the label

function onHoverCallback() {
    // Create a label here
    // Example: Show a label at the cursor position
    ligandLabel = viewer.addLabel("O08 JGY 403 D", {
        position: {x: 4.64, y: -0.155, z: -9.262},
        alignment: "center", borderColor: "black",
        borderThickness: 2, fontSize: 24, 
        backgroundColor: 'white',
        fontColor: "black",
    });
    interLabel = viewer.addLabel("Weak Polar 2.96Å", {
        position: {x: 4.755, y: 1.079, z: -10.067},
        alignment: "center", borderColor: "#fc7600",
        borderThickness: 2, fontSize: 24, 
        backgroundColor: 'white',
        fontColor: "#fc7600",
    });
    protLabel = viewer.addLabel("CB GLN 299 A", {
        position: {x: 4.869, y: 2.313, z: -10.873},
        alignment: "center", borderColor: "black",
        borderThickness: 2, fontSize: 24, 
        backgroundColor: 'white',
        fontColor: "black",
    });
}

function onUnhoverCallback() {
    // Use the shared variable to remove or hide the label
    if (ligandLabel && interLabel && protLabel) {
        viewer.removeLabel(ligandLabel);
        viewer.removeLabel(interLabel);
        viewer.removeLabel(protLabel);
        ligandLabel = null; // Reset the reference
        interLabel = null; // Reset the reference
        protLabel = null; // Reset the reference
        viewer.render(); // Re-render to update the view
    }
}

viewer.addCylinder(
    {
        start:{x: 4.64, y: -0.155, z: -9.262 },
        end:{x: 4.869, y: 2.313, z: -10.873 },
        radius: 0.125, dashed: true, fromCap:1, toCap:1, color:"#fc7600",
        hoverable: true, hover_callback: onHoverCallback,
        unhover_callback: onUnhoverCallback,
    }
)

viewer.render();
```

This works fine and label is displayed on hover and hidden (removed) when I hover out of it.

I am now trying to extend this behaviour to any number of cylinder shapes.

This is the code I have:
```
function onHoverCallback(contactBgnCoords, contactEndCoords, contactColor,
    contactDistance, proteinChainId, ligandChainId,
    proteinResNum, ligandResNum, proteinResName, ligandResName,
    proteinAtomName, ligandAtomName, theContactType) {

    var middlePoint = findMiddlePoint(contactBgnCoords, contactEndCoords);
    // Create a label here
    // Example: Show a label at the cursor position
    ligandLabel = viewer.addLabel(`${ligandAtomName} ${ligandResName} ${ligandResNum} ${ligandChainId}`, {
        position: {x: contactBgnCoords[0], y: contactBgnCoords[1], z: contactBgnCoords[2]},
        alignment: "center", borderColor: "black",
        borderThickness: 2, fontSize: 12, 
        backgroundColor: 'white',
        fontColor: "black",
    });
    interLabel = viewer.addLabel(`${theContactType} ${contactDistance}Å`, {
        position: {x: middlePoint[0], y: middlePoint[1], z: middlePoint[2]},
        alignment: "center", borderColor: contactColor,
        borderThickness: 2, fontSize: 12, 
        backgroundColor: 'white',
        fontColor: contactColor,
    });
    protLabel = viewer.addLabel(`${proteinAtomName} ${proteinResName} ${proteinResNum} ${proteinChainId}`, {
        position: {x: contactEndCoords[0], y: contactEndCoords[1], z: contactEndCoords[2]},
        alignment: "center", borderColor: "black",
        borderThickness: 2, fontSize: 12, 
        backgroundColor: 'white',
        fontColor: "black",
    });

    cylinderLabels.push([ligandLabel, interLabel, protLabel]);

}

function onUnhoverCallback(index) {
    // remove labels and empty index element of cylinderLabels
    viewer.removeLabel(cylinderLabels[index][0]); // Remove the ligand atom label
    viewer.removeLabel(cylinderLabels[index][1]); // Remove the interaction type label
    viewer.removeLabel(cylinderLabels[index][2]); // Remove the protein atom label
    //cylinderLabels[index] = [];
    cylinderLabels[index][0] = null;
    cylinderLabels[index][1] = null;
    cylinderLabels[index][2] = null;
    viewer.render();
}

// Running the viewer.addCylinder command
                var contactCylinder = viewer.addCylinder({
                    start: { x: contactBgnCoords[0], y: contactBgnCoords[1], z: contactBgnCoords[2] },
                    end: { x: contactEndCoords[0], y: contactEndCoords[1], z: contactEndCoords[2] },
                    radius: contactWidth,
                    dashed: true,
                    fromCap: 1,
                    toCap: 1,
                    color: contactColor,
                    hoverable: true,
                    hover_callback: onHoverCallback(
                        contactBgnCoords, contactEndCoords, contactColor,
                        contactDistance, proteinChainId, ligandChainId,
                        proteinResNum, ligandResNum, proteinResName, ligandResName,
                        proteinAtomName, ligandAtomName, theContactType,
                        ),
                    unhover_callback: onUnhoverCallback(index),
                });
contactCylinders.push(contactCylinder);
```

This is not really working. It seems every time a label is added is instantly removed, calling hover and unhover callback. By the time the viewer is rendered the labels are not there (this is as it should be), but nothing happens when I hover on them and when hover out of them I start getting the same error many times: 

<img width="703" alt="Screenshot 2024-08-19 at 16 39 33" src="https://github.com/user-attachments/assets/bbd76e1f-bb8a-4ba9-afb9-f9876cf5f45f">

Is this a possible behaviour? The issue might be my `JavaScript` code as I am not an expert. The labels are correctly rendered and then un-rendered, but then they don't work, neither on hover or when hovering out.

Help with this would be much appreciated. Many thanks!

**Comment by dkoes** [comment]
hover_callback and unhover_callback need to be functions, but you are setting them to the return value of these functions (which is nothing). 

Your code would be cleaner if you made use of the fact that the callbacks are passed the selected object.  You can store the labels in the object itself and not have to have a global variable.

**Comment by JavierSanchez-Utges** [comment]
Hi, thanks, David! Interesting, I thought I had commented on this yesterday after I figured out. After reading some other somewhat related issues and checking the code of some of your examples I figured it out exactly like you said on your comment! I defined the function right there on the `addCylinder` spec. I used the `this` (Cylinder shape) object passed to the function where I had stored all the metadata needed for the labels in `userData`. I accessed it using `this.stylespec.userData`

<img width="250" alt="Screenshot 2024-08-20 at 07 28 41" src="https://github.com/user-attachments/assets/ecae67b0-6be3-443a-9fdd-b9abfa8775f4">

<img width="250" alt="Screenshot 2024-08-20 at 07 29 20" src="https://github.com/user-attachments/assets/4ffe5ecb-4b55-4215-a43c-ce59358a4d03">

<img width="250" alt="Screenshot 2024-08-20 at 07 30 06" src="https://github.com/user-attachments/assets/6c5b0a44-83c2-4f3b-a95c-3ef135d309b0">

Thanks a lot!
